mlflow/tensorflow/__init__.py

```diff
@@ -495,31 +525,46 @@ def save_model(
     _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))
 
 
+def _load_custom_objects(path, file_name):
+    custom_objects_path = None
+    if os.path.isdir(path):
+        if os.path.isfile(os.path.join(path, file_name)):
+            custom_objects_path = os.path.join(path, file_name)
+    if custom_objects_path is not None:
+        import cloudpickle
+
+        with open(custom_objects_path, "rb") as f:
+            return cloudpickle.load(f)
+
+
 def _load_keras_model(model_path, keras_module, save_format, **kwargs):
     keras_models = importlib.import_module(keras_module.__name__ + ".models")
     custom_objects = kwargs.pop("custom_objects", {})
-    custom_objects_path = None
+    if saved_custom_objects := _load_custom_objects(model_path, _CUSTOM_OBJECTS_SAVE_PATH):
+        saved_custom_objects.update(custom_objects)
+        custom_objects = saved_custom_objects
+
+    if global_custom_objects := _load_custom_objects(model_path, _GLOBAL_CUSTOM_OBJECTS_SAVE_PATH):
+        global_custom_objects.update(custom_objects)
+        custom_objects = global_custom_objects
+
     if os.path.isdir(model_path):
-        if os.path.isfile(os.path.join(model_path, _CUSTOM_OBJECTS_SAVE_PATH)):
-            custom_objects_path = os.path.join(model_path, _CUSTOM_OBJECTS_SAVE_PATH)
         model_path = os.path.join(model_path, _MODEL_SAVE_PATH)
-    if custom_objects_path is not None:
-        import cloudpickle
-
-        with open(custom_objects_path, "rb") as in_f:
-            pickled_custom_objects = cloudpickle.load(in_f)
-            pickled_custom_objects.update(custom_objects)
-            custom_objects = pickled_custom_objects
 
     # If the save_format is HDF5, then we save with h5 file
     # extension to align with prior behavior of mlflow logging
     if save_format == "h5":
-        model_path = model_path + ".h5"
+        model_path += ".h5"
+    # Since TF 2.16.0, it only supports saving model in .h5 or .keras format.
+    # But for backwards compatibility, we still save model without suffix
+    # for older versions of TF.
+    elif os.path.exists(model_path + ".keras"):
+        model_path += ".keras"
 
     # keras in tensorflow used to have a '-tf' suffix in the version:
     # https://github.com/tensorflow/tensorflow/blob/v2.2.1/tensorflow/python/keras/__init__.py#L36
     unsuffixed_version = re.sub(r"-tf$", "", _get_keras_version(keras_module))
-    if save_format == "h5" and Version(unsuffixed_version) >= Version("2.2.3"):
+    if save_format == "h5" and Version("2.2.3") <= Version(unsuffixed_version) <= Version("2.15.0"):
```

Comment:
> maybe we should just update them if 2.15.1 is out

this is annoying.

I think we can use `Version(keras.__version__).release < (2, 16)` instead, it will remove "dev" firstly, then do comparison.